lib/srv/db/cloud/users/users.go

```diff
@@ -0,0 +1,285 @@
+func (u *Users) setupDatabase(ctx context.Context, database types.Database) {
+	// Database metadata is already refreshed once during database
+	// registration so no need to do it again.
+	u.setupDatabases(ctx, types.Databases{database}, false /*updateMeta*/)
+}
+
+// setupAllDatabases performs setup for all databases.
+func (u *Users) setupAllDatabases(ctx context.Context, allDatabases types.Databases) {
+	u.setupDatabases(ctx, allDatabases, true)
+
+	// Clean up.
+	u.lookup.removeUnusedDatabases(allDatabases)
+	activeUsers := u.lookup.usersByID()
+	for userID, user := range u.usersByID {
+		if _, found := activeUsers[userID]; !found {
+			delete(u.usersByID, userID)
+
+			if err := user.Teardown(ctx); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to tear down user %v.", user)
+			}
+		}
+	}
+}
+
+// setupDatabases performs setup for provided databases.
+func (u *Users) setupDatabases(ctx context.Context, databases types.Databases, updateMeta bool) {
+	for _, database := range databases {
+		fetcher, found := u.fetchersByType[database.GetType()]
+		if !found {
+			continue
+		}
+
+		// Refresh database metadata like ElastiCache user group IDs.
+		//
+		// Auto discovered databases can be skipped as the watcher service
+		// discovers changes periodically then triggers reconciler.
+		//
+		// If UpdateMeta fails, log an error and continue to next step with
+		// whatever meta database already has.
+		if updateMeta && database.Origin() != types.OriginCloud {
+			if err := u.cfg.UpdateMeta(ctx, database); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to update metadata for %q.", database)
+			}
+		}
+
+		// Fetch users.
+		fetchedUsers, err := fetcher.FetchDatabaseUsers(ctx, database)
+		if err != nil {
+			if trace.IsAccessDenied(err) { // Permission errors are expected.
+				u.cfg.Log.WithError(err).Debugf("No permissions to fetch users for %q.", database)
+			} else {
+				u.cfg.Log.WithError(err).Errorf("Failed to fetch users for database %v.", database)
+			}
+			continue
+		}
+
+		// Setup users.
+		var users []User
+		for _, fetchedUser := range fetchedUsers {
+			if user, err := u.setupUser(ctx, fetchedUser); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to setup user %v for database %v.", fetchedUser, database)
+			} else {
+				users = append(users, user)
+			}
+		}
+
+		// Rotate passwords.
+		for _, user := range users {
+			if err = user.RotatePassword(ctx); err != nil {
+				u.cfg.Log.WithError(err).Errorf("Failed to rotate password for user %v", user)
+			}
+		}
+
+		// Update lookup.
+		u.lookup.setDatabaseUsers(database, users)
+	}
+}
+
+// setupUser finds existing user if it is already managed and tracked,
+// otherwise try to setup the new user.
+func (u *Users) setupUser(ctx context.Context, user User) (User, error) {
+	if existingUser, found := u.usersByID[user.GetID()]; found {
+		// TODO(greedy52) may want to compare secret store setting in case they
```

Comment:
The secret store setting is set per database in file config and databases can share the same user. The behaviour will be undefined if they are different. It's probably ok if different KMS keys, but different secret prefixes can be catastrophic. Even worse if configs are different across db agents in HA mode.

We did ask the product team on this. Instead of trying to minimize the impact in the code, we will for now just ask the users to be aware and be consistent when they are configuring the service.